### ISSUE TITLE
loop optimization + pytorch (028-xxx)

### ISSUE THREAD

**Issue by yzbx** [ISCAS007/torchseg#31]
# focal_loss028/029
- **gamma = 0.5 train failed**
- gamma = 1.0 miou=0.377 (for upsamle_layer=3) 0.386 (for upsample_layer=5)
- gamma = 2.0 is slight better than gamma=1.0

```
python test/pspnet_test.py --backbone_pretrained=True --backbone_name=resnet50 \
--upsample_type=bilinear --midnet_scale=15 --batch_size=4 --note=focal_loss028 \
--test=hyperopt --hyperkey=model.focal_loss_gamma --hyperopt_calls=6 \
--use_momentum=False --upsample_layer=3 --dataset_use_part=320

python test/pspnet_test.py --backbone_pretrained=True --backbone_name=resnet50 \
--upsample_type=bilinear --midnet_scale=15 --batch_size=2 --note=focal_loss029 \
--test=hyperopt --hyperkey=model.focal_loss_gamma --hyperopt_calls=6 \
--use_momentum=True --upsample_layer=5 --dataset_use_part=320
```

|                        |   train/iou |   val/iou |   n_epoch |
|:-----------------------|------------:|----------:|----------:|
| ('focal_loss028', 0.5) |    0.01994  | 0.0192171 |       100 |
| ('focal_loss028', 1.0) |    0.570953 | 0.377005  |       100 |
| ('focal_loss028', 2.0) |    0.556718 | 0.378915  |       100 |

|                        |   ('val/iou', 'mean') |   ('val/iou', 'std') |   ('val/iou', 'amax') |
|:-----------------------|----------------------:|---------------------:|----------------------:|
| ('focal_loss028', 0.5) |             0.0192171 |           0          |             0.0192171 |
| ('focal_loss028', 1.0) |             0.377005  |           0.00549323 |             0.38089   |
| ('focal_loss028', 2.0) |             0.378915  |           0.0059462  |             0.38312   |

|    |   train/iou |   val/iou | note          |   focal_loss_gamma |   n_epoch |
|---:|------------:|----------:|:--------------|-------------------:|----------:|
|  2 |   0.0191035 | 0.0192171 | focal_loss028 |                0.5 |       100 |
|  5 |   0.0207766 | 0.0192171 | focal_loss028 |                0.5 |       100 |
|  4 |   0.575547  | 0.373121  | focal_loss028 |                1   |       100 |
|  0 |   0.562847  | 0.374711  | focal_loss028 |                2   |       100 |
|  3 |   0.566359  | 0.38089   | focal_loss028 |                1   |       100 |
|  1 |   0.550588  | 0.38312   | focal_loss028 |                2   |       100 |

|                        |   train/iou |   val/iou |   n_epoch |
|:-----------------------|------------:|----------:|----------:|
| ('focal_loss029', 0.5) |   0.0191787 | 0.0192171 |       100 |
| ('focal_loss029', 1.0) |   0.635259  | 0.386365  |       100 |
| ('focal_loss029', 2.0) |   0.624699  | 0.403327  |       100 |

|                        |   ('val/iou', 'mean') |   ('val/iou', 'std') |   ('val/iou', 'amax') |
|:-----------------------|----------------------:|---------------------:|----------------------:|
| ('focal_loss029', 0.5) |             0.0192171 |           0          |             0.0192171 |
| ('focal_loss029', 1.0) |             0.386365  |           0.00394113 |             0.389151  |
| ('focal_loss029', 2.0) |             0.403327  |           0.00978559 |             0.410246  |

|    |   train/iou |   val/iou | note          |   focal_loss_gamma |   n_epoch |
|---:|------------:|----------:|:--------------|-------------------:|----------:|
|  0 |   0.0188437 | 0.0192171 | focal_loss029 |                0.5 |       100 |
|  2 |   0.0195137 | 0.0192171 | focal_loss029 |                0.5 |       100 |
|  1 |   0.629185  | 0.383578  | focal_loss029 |                1   |       100 |
|  5 |   0.641334  | 0.389151  | focal_loss029 |                1   |       100 |
|  3 |   0.616328  | 0.396407  | focal_loss029 |                2   |       100 |
|  4 |   0.63307   | 0.410246  | focal_loss029 |                2   |       100 |

**Comment by yzbx** [comment]
# norm_ways30
```
for times in 1 2 3 4 5
do
    for norm_ways in caffe pytorch cityscapes -1,1 0,1
    do
        python test/pspnet_test.py --batch_size=2 \
        --backbone_pretrained=True --midnet_scale=15 \
        --upsample_type=bilinear --dataset_use_part=320 \
        --note=norm_ways030
    done
done
```

**Comment by dishen12** [comment]
# fl_mult031 (deeplab2)
```
python test/pspnet_test.py --test=hyperopt --n_epoch=50 --note=fl_mult031 --upsample_type=bilinear --backbone_pretrained=True --midnet_scale=5 --dataset_use_part=320 --focal_loss_gamma=2.0 --changed_lr_mult=2 --new_lr_mult=5 --use_momentum=True --upsample_layer=4 --batch_size=32 --hyperkey=model.changed_lr_mult --hyperopt_calls=8
```

**Comment by yzbx** [comment]
# fl_alpha32
**focal loss alpha in [1.0, 5.0 ,10.0]
- [ ] fl_alpha32
```
python test/pspnet_test.py --test=hyperopt --n_epoch=50 \
--note=fl_alpha032 --upsample_type=bilinear \
--backbone_pretrained=True --midnet_scale=15 \
--dataset_use_part=320 --focal_loss_gamma=2.0 \
--changed_lr_mult=2 --new_lr_mult=10 --use_momentum=False \
--upsample_layer=3 --batch_size=6 \
--hyperkey=model.focal_loss_alpha --hyperopt_calls=6
```

**Comment by yzbx** [comment]
# batch size (dl)
```
for batch_size in 8 16 32
do
    python test/pspnet_test.py --batch_size=${batch_size} \
    --backbone_pretrained=True --midnet_scale=5 \
    --backbone_freeze=False --backbone_name=vgg16 \
    --upsample_type=bilinear --dataset_use_part=320 \
    --note=bs${batch_size}
done
```

**Comment by yzbx** [comment]
# use_dropout034
**config->get_midnet->os.environ->model**
```
...
```

**Comment by yzbx** [comment]
# use_bias035 (dl)
```
python test/pspnet_test.py --use_dropout False --use_lr_mult False --dataset_use_part=320 \
--backbone_pretrained=True --midnet_scale=15 --n_epoch=50 --upsample_type=bilinear \
--hyperopt_calls=6 --hyperkey=model.use_bias --note=use_bias035 --test=hyperopt
```

**Comment by dishen12** [comment]
# moment_bias036 (deeplab2 118)
momentum in [0.1, 0.05, 0.01]
bias in [True False]
```
python test/pspnet_test.py --use_dropout False --use_lr_mult False --dataset_use_part=320 \
--backbone_pretrained=True --midnet_scale=15 --n_epoch=50 --upsample_type=bilinear \
--hyperopt_calls=12 --hyperkey=model.momentum,model.use_bias --note=moment_bias036 --test=hyperopt
```

**Comment by dishen12** [comment]
# lr_mult037 (deeplab3 237)
'model.changed_lr_mult':('choices',[1,2,5]),
'model.new_lr_mult':('choices',[1,5,10]),

```
python test/pspnet_test.py --use_dropout False --use_lr_mult True --dataset_use_part=320 \
--backbone_pretrained=True --midnet_scale=10 --n_epoch=50 --upsample_type=bilinear \
--hyperopt_calls=27 --hyperkey=model.changed_lr_mult,model.new_lr_mult--note=lr_mult037\
--test=hyperopt --batch_size=16
```

**Comment by yzbx** [comment]
# optimizer038 (dl)
adam series optimizer benchmark
```
python test/pspnet_test.py --test=hyperopt --use_lr_mult=False --midnet_scale=10 --batch_size=4 --learning_rate=1e-4 --hyperopt=loop --hyperkey=model.optimizer --hyperopt_calls=9 --momentum=0.01 --note=optimizer038
```

**Comment by yzbx** [comment]
# upsample_layer039 vs upsample_layer023
backbone_name resnet50 vs vgg16
midnet_scale 10 vs 15 
dataset_use_part 640 vs 320
use_lr_mult=True

- [ ] code
```
python test/pspnet_test.py --test=hyperopt --use_lr_mult=True --changed_lr_mult=2 \
--new_lr_mult=10 --midnet_scale=10 --dataset_use_part=640 \
--batch_size=4 --learning_rate=1e-4 --hyperopt=loop --hyperkey=model.upsample_layer \
--hyperopt_calls=6 use_momentum=True --momentum=0.01 --note=upsample_layer039 
```

**Comment by yzbx** [comment]
# batch_size040

- [ ] running
```
python test/pspnet_test.py --test=hyperopt --use_lr_mult=True --changed_lr_mult=2 \
--new_lr_mult=10 --midnet_scale=10 --upsample_layer=4 --dataset_use_part=320 \
--batch_size=4 --learning_rate=1e-4 --hyperopt=loop --hyperkey=args.batch_size \
--hyperopt_calls=6 use_momentum=True --momentum=0.01 --note=batch_size040 
```

**Comment by yzbx** [comment]
# res50_fz_ul041 + vgg16_fz_ul041
* backbone_name [resnet50, vgg16]
* upsample_layer [3,4,5]

note: **--backbone_freeze=True**

```
python test/pspnet_test.py --midnet_scale=10 --backbone_name=resnet50 --backbone_freeze=True \
--batch_size=2 --use_momentum=True --momentum=0.01 --changed_lr_mult=1 --new_lr_mult=2 \
--use_lr_mult=True --note=res50_fz_ul041 --upsample_layer=5 --test=hyperopt \
--hyperkey=model.upsample_layer --hyperopt_calls=6

python test/pspnet_test.py --midnet_scale=10 --backbone_name=vgg16 --backbone_freeze=True \
--batch_size=2 --use_momentum=True --momentum=0.01 --changed_lr_mult=1 --new_lr_mult=2 \
--use_lr_mult=True --note=vgg16_fz_ul041 --upsample_layer=5 --test=hyperopt \
--hyperkey=model.upsample_layer --hyperopt_calls=6
```

**Comment by yzbx** [comment]
# fl_grad042
```
python test/pspnet_test.py --midnet_scale=10 --backbone_name=resnet50 --batch_size=4 --use_lr_mult=False --note=fl_grad042 --upsample_layer=3 --test=hyperopt --hyperkey=model.focal_loss_grad --focal_loss_gamma=2.0 --hyperopt_calls=6 --dataset_use_part=320
```

**Comment by yzbx** [comment]
# scheduler043 (todo)

```
python test/pspnet_test.py --midnet_scale=10 --backbone_name=vgg16 --batch_size=2 --use_lr_mult=False --note=scheduler043_poly --upsample_layer=4 --dataset_use_part=320 --learning_rate=1e-2 --optimizer=sgd

python test/pspnet_test.py --midnet_scale=10 --backbone_name=vgg16 --batch_size=2 --use_lr_mult=False --note=scheduler043_adam --upsample_layer=4 --dataset_use_part=320 --learning_rate=1e-4 --optimizer=adam

python test/pspnet_test.py --midnet_scale=10 --backbone_name=vgg16 --batch_size=2 --use_lr_mult=False --note=scheduler043_pop --upsample_layer=4 --dataset_use_part=320 --learning_rate=1e-2 --optimizer=sgd --scheduler=poly_pop
```

**Comment by yzbx** [comment]
# resnet_nw044 
python test/pspnet_test.py --midnet_scale=10 --backbone_name=resnet50--batch_size=2 --use_lr_mult=False --note=scheduler043_pop --upsample_layer=4 --dataset_use_part=320 ----test=hyperopt --hyperkey=dataset.norm_ways--hyperopt_calls=1 --n_epoch=50

**Comment by yzbx** [comment]
# freeze_layer045
python test/pspnet_test.py --midnet_scale=10 --backbone_name=resnet50 --batch_size=2 --use_lr_mult=False --note=freeze_layer045--upsample_layer=4 --dataset_use_part=320 --test=hyperopt --hyperkey=model.freeze_layer --hyperopt_calls=2 --n_epoch=50 --backbone_freeze=False --modify_resnet_head=False

**Comment by yzbx** [comment]
# crop_size_step046
python test/pspnet_test.py --backbone_name=vgg16 --backbone_freeze=True --n_epoch=50 --keep_crop_ratio=False --note=crop_size_step046 --min_crop_size=720 --max_crop_size=1024 --test=hyperopt --hyperkey=aug.crop_size_step --hyperopt_calls=2 --dataset_use_part=320

**Comment by yzbx** [comment]
# crop_size_step047 
python test/pspnet_test.py --backbone_name=vgg16 --backbone_freeze=True --n_epoch=50 --keep_crop_ratio=False --note=crop_size_step047 --min_crop_size=480 --max_crop_size=1080 --test=hyperopt --hyperkey=aug.crop_size_step --hyperopt_calls=2 --dataset_use_part=320

**Comment by dishen12** [comment]
# freeze_layer048 
python test/pspnet_test.py --midnet_scale=10 --backbone_name=resnet50 --batch_size=2 --use_lr_mult=False --note=freeze_layer048 --upsample_layer=5 --dataset_use_part=320 --test=hyperopt --hyperkey=model.freeze_layer --hyperopt_calls=2 --n_epoch=50 --backbone_freeze=False --modify_resnet_head=False

**Comment by dishen12** [comment]
# pad_for_crop049
python test/pspnet_test.py --backbone_name=vgg19_bn --backbone_freeze=True --n_epoch=50 --keep_crop_ratio=False --note=pad_for_crop049 --min_crop_size=480 --max_crop_size=1080 --test=hyperopt --hyperkey=aug.crop_size_step --hyperopt_calls=2 --dataset_use_part=320 --pad_for_crop=True

**Comment by dishen12** [comment]
# momentum050
python test/pspnet_test.py --use_dropout False --use_lr_mult False --dataset_use_part=320 --backbone_pretrained=True --midnet_scale=5 --n_epoch=50 --upsample_type=bilinear --hyperopt_calls=3 --use_momentum=True --hyperkey=model.momentum --note=moment050 --test=hyperopt --batch_size=16

**Comment by dishen12** [comment]
# freeze_ratio051
python test/pspnet_test.py --backbone_name=resnet50 --use_lr_mult=True --freeze_ratio=0.5 --upsample_layer=5 --modify_resnet_head=False --n_epoch=50 --dataset_use_part=320 --test=hyperopt --hyperkey=model.freeze_ratio --hyperopt_calls=2 --note=freeze_ratio051

python test/pspnet_test.py --backbone_name=vgg19_bn --use_lr_mult=True --freeze_ratio=0.5 --upsample_layer=5 --modify_resnet_head=False --n_epoch=50 --dataset_use_part=320 --test=hyperopt --hyperkey=model.freeze_ratio --hyperopt_calls=2 --note=freeze_ratio051

**Comment by yzbx** [comment]
# batch norm vs group norm
## vgg16 baseline **0.589**
python test/pspnet_test.py --backbone_name=vgg16_bn --use_momentum=True --midnet_scale=10 --upsample_layer=4 --use_lr_mult=False --batch_size=4 --note=vgg16_bn

## vgg16 + duc **0.623**
python test/pspnet_test.py --backbone_name=vgg16_bn --use_momentum=True --midnet_scale=10 --upsample_layer=4 --use_lr_mult=False --batch_size=4 --upsample_type=duc --note=vgg16_duc

## vgg16 + gn **0.5802**
python test/pspnet_test.py --backbone_name=vgg16_gn --use_momentum=True --midnet_scale=10 --upsample_layer=4 --use_lr_mult=False --batch_size=4 --note=vgg16_gn

**Comment by yzbx** [comment]
# upsample_type053
python test/pspnet_test.py --backbone_name=vgg16_gn --use_lr_mult=False --upsample_layer=4 --batch_size=4 --use_momentum=True --n_epoch=50 --dataset_use_part=320 --test=hyperopt --hyperkey=model.upsample_type --hyperopt_calls=2 --note=upsample_type053
